Nip 94 - File Header (#337)
nostr-protocol/nips: 94.md

NIP-94 - File Header
======
`draft` `optional` `author:frbitten` 

The purpose of this NIP is to allow an organization and classification of shared files. So that relays can filter and organize in any way that is of interest.
Also the goal is to create a base on the protocol for this bountsr "Filesharing App" (https://bountsr.org/p2p-filesharing/) to be implemented.

Nostr event
------------------
This NIP specifies the use of the `30063` event type (parameterized replaceable event), having in `content` a description of the file content, and a list of tags described below:
* `d` containing the name of the shared file to allow it to be overwritten in the future.
* `url` the url to download the file
* `type` a string indicating the data type of the file. The MIME types format must be used (https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types)

> * `decrypt` (optional) If the file is encrypted, you must indicate in the first value the algorithm used to encrypt the file and in the second value the information (the key) that allow the decryption of the file

Reviewer: `Key` is probably too specific here, Maybe more generically `Decryption Params` because not every encryption scheme will use the same params

Reviewer: Ref: https://github.com/bitwiseshiftleft/sjcl/blob/master/core/convenience.js#L14

NIP-94 - File Header
======
`draft` `optional` `author:frbitten` 

> The purpose of this NIP is to allow an organization and classification of shared files. So that relays can filter and organize in any way that is of interest.

Reviewer 2: immutable shared files?

Also the goal is to create a base on the protocol for this bountsr "Filesharing App" (https://bountsr.org/p2p-filesharing/) to be implemented.

Nostr event
------------------

> This NIP specifies the use of the `1063` event type (parameterized replaceable event), having in `content` a description of the file content, and a list of tags described below:

Reviewer 2: It's not a "parameterized replaceable event" right?

Author: Yes. According to some suggestions, being a replaceable event would lead to problems and difficulties in keeping relationships up to date. And if it is a new URI it is a new file that must be used a new event.

> * `url` the url to download the file

Reviewer 3: `r` is already used elsewhere as "reference" with url value.

The reason to make it searchable is that I imagined a client may want to have an user media panel that the user would fill by using a context menu pop-up over images on notes (like "add to memes collection" for instance).

For that, it would detect used image url on kind 1 note content and create a corresponding kind 1063 event. Making 1063 url tag searchable will allow the client to know if an event with that url from the same pubkey already exists.

It would be an alternative to searching by hash.

```suggestion
* `r` the url to download the file, without trailing slash
```

> * `type` a string indicating the data type of the file. The MIME types format must be used (https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types)

Reviewer 4: It might be nice if this was searchable. I don't want to load exe files, but I do want to load pngs.

> * `decrypt` (optional) If the file is encrypted, you must indicate in the first value the algorithm used to encrypt the file and in the second value the parameters that allow the decryption of the file

Reviewer 5: ```suggestion
* `xchacha20` (optional) If the file is encrypted using XChaCha20, you must indicate in the first value the encryption key and in the second value the nonce, as hex
```

Reviewer: Are there decent libs for `xchacha20` ? void.cat uses AES-GCM

Reviewer 5: I'm happy with AES-GCM. I just heard XChaCha20 is "better". I just think we need to decide on one.

Reviewer: XChaCha20 is much newer so it's a bit harder to find decent libs, while AES ubiquitous

> * `decrypt` (optional) If the file is encrypted, you must indicate in the first value the algorithm used to encrypt the file and in the second value the parameters that allow the decryption of the file

Reviewer 6: ```suggestion

```
Why would we want public files to be encrypted?
I propose this, although optional, can be discarded completely in favor of simplicity.

Author: The idea of encryption in public files is that the server hosting the file does not know its contents. Thus avoiding censorship on the hosting server. It is a suggestion from NIP-68 (#250 ) that I have incorporated into the proposal.

> * `hash` containing the SHA-256 hexencoded string of the file or the BlurHash string (for images). The third array value is the lowercase chosen algorithm name: "sha256" or "blurhash".

Reviewer 3: Just ocurred to me that `h` instead of `hash` will make it possible to find by the tag, so to know if a file header already exists before effectively uploading a file.

```suggestion
* `h` containing the SHA-256 hexencoded string of the file or the BlurHash string (for images). The third array value is the lowercase chosen algorithm name: "sha256" or "blurhash".
```

Reviewer 5: I think the best way would be to replace this with two tags: `h`, which contains the hex-encoded lowercase sha256 hash of the file, for search purposes (our own little IPFS), and `blurhash` in a separate tag, as it has just cosmetic purposes.

Reviewer 5: Why not make it `h` so it is searchable? I think that, since users can write anything in the `sha256` field and not necessarily the correct hash, we will need relays to volunteer support for this extra requirement, if they want, and block events that have the wrong hashes, otherwise the system will not work.

If relays are already going through this tedious and intensive work of downloading the files and validating their hashes, they may as well support queries for the filter `#sha256`.

```json
{
  "id": <32-bytes lowercase hex-encoded sha256 of the the serialized event data>,
  "pubkey": <32-bytes lowercase hex-encoded public key of the event creator>,
  "created_at": <unix timestamp in seconds>,
  "kind": 1063,
  "tags": [

>     ["url",<string with URL of file>],

Reviewer 3: `r` instead of `url` to make it searchable

```suggestion
    ["r", <string with URL of file>],
```

Reviewer 6: ```suggestion
    ["r",<string with URI of file>],
```
@arthurfranca note I'm using `URI` instead of a `url` since this reference should be an identifier, not necessarily a location.
I suggest using a general uri to accommondate different ways of receiving data.

>     ["decrypt",<algorithm>,<Decryption Params>],

Reviewer 5: ```suggestion
    ["xchacha20", <key>, <nonce>],
```

Reviewer: `["aes-256-gcm", "<key>", "<iv>", "<ts>"]`

Where `key` & `iv` are hex, `ts` is the Tagsize in bits

Reviewer: https://github.com/bitwiseshiftleft/sjcl/blob/85caa53c281eeeb502310013312c775d35fe0867/core/gcm.js#L17-L24

Reviewer: Or you could specify the tagsize is 128 its and just leave it out of the tag

    ["type", <MIME type>],

>     ["hash", "LKN]Rv%2Tw=w]~RBVZRi};RPxuwH", <"blurhash" or "sha256">]

Reviewer 3: `h` instead of `hash` to make it searchable
```suggestion
    ["h", "LKN]Rv%2Tw=w]~RBVZRi};RPxuwH", <"blurhash" or "sha256">]
```

Author: I didn't put it as a letter because I don't see why the Hash tag should be indexed. I see no reason to do a database hash lookup.
